### PR TITLE
websocket support

### PR DESCRIPTION
**Pull request by trbngr**
Added WebSocket subscriptions. The resulting code is pretty straight forward. Let me know if you have any questions and/or improvements.

### CODE REVIEW

File: src/main/scala/Server.scala [hunk 1]
```diff
@@ -102,46 +70,44 @@ object Server extends App {
         }
 
       case Failure(error: SyntaxError) ⇒
-        complete(BadRequest, JsObject(
+        complete(ToResponseMarshallable(BadRequest → JsObject(
```
Comment (trbngr) [inline src/main/scala/Server.scala:73]: Not sure if calling ToResponseMarshallable is necessary. I made the change just to make the compiler happy.

File: src/main/scala/Server.scala [hunk 2]
```diff
@@ -102,46 +70,44 @@ object Server extends App {
...
+            case Some(obj: JsObject) ⇒ obj
+            case _ ⇒ JsObject.empty
+          }
 
-        executeQuery(query, operation, vars)
-      }
+          executeQuery(query, operation, vars)
+        }
+      } ~
+        get(handleWebSocketMessages(graphQlSubscriptionSocket(subscriptionEventPublisher, ctx)))
```
Comment (trbngr) [inline src/main/scala/Server.scala:103]: Nice and clean WS integration.

File: src/main/scala/SubscriptionActor.scala [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+      publisher ! SubscriptionEventPublisher.Join
+      context.become(connected(outgoing))
+  }
+
+  def connected(outgoing: ActorRef): Receive = {
+    case e: Subscribe =>
+      log.info(s"Got sub: $e")
+      prepareQuery(e)
+
+    case context: PreparedQueryContext =>
```
Comment (trbngr) [inline src/main/scala/SubscriptionActor.scala:43]: I'm using a wrapper class. In my project, I need to pass around the variables with the subscription.

**Comment by trbngr** [comment]
This also allows for multiple subscriptions per socket.

File: src/main/scala/SubscriptionActor.scala [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+import generic.Event
+import sangria.ast.OperationType
+import sangria.execution.{Executor, PreparedQuery}
+import sangria.marshalling.sprayJson._
+import sangria.parser.QueryParser
+import spray.json._
+
+import scala.util.{Failure, Success}
+
+object SubscriptionActor extends DefaultJsonProtocol {
```
Comment (trbngr) [inline src/main/scala/SubscriptionActor.scala:12]: Maybe a better name is `SubscriptionClientActor`?

File: src/main/scala/Server.scala (hunk 1 above)
Comment (OlegIlyenko) [inline src/main/scala/Server.scala:73]: I also had this issue before. Type inference is unable to figure out the types, so I also did it explicitly, just like you did here.

File: src/main/scala/SubscriptionActor.scala [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  }
+
+  def prepareQuery(subscription: Subscribe) = {
+    QueryParser.parse(subscription.query) match {
+      case Success(ast) =>
+        ast.operationType(subscription.operation) match {
+          case Some(OperationType.Subscription) =>
+            executor.prepare(ast, ctx, (), subscription.operation, JsObject.empty).map {
+              query => self ! PreparedQueryContext(query)
+            }
```
Comment (OlegIlyenko) [inline src/main/scala/SubscriptionActor.scala:74]: You don't handle an error here. For instance, if query contains validation errors, the error would be just swallowed without any logs or feedback to a client.

File: src/main/scala/SubscriptionActor.scala [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+          case _ => Set(context)
+        })
+      }
+
+    case event: Event =>
+      val fieldName = schema.subscriptionFieldName(event)
+      queryContextsFor(fieldName) foreach { ctx =>
+        ctx.query.execute(root = event) map { result =>
+          outgoing ! QueryResult(result)
+        }
```
Comment (OlegIlyenko) [inline src/main/scala/SubscriptionActor.scala:58]: It's the same issue with error handling here (as in the `executor.prepare` code below). I understand that this code is just an example/demonstration of the concept, so some of the details are skipped, Still I think error handling is important for these 2 operations and for the demonstration itself. Example code is very popular source of copy/paste, so if people copy/past then it's better be with error handling :)

**Comment by OlegIlyenko** [comment]
Just reviewed the code. I really like the websoket implementation! :+1: 

I'm not sure about merging it as it is though. The thing is: I view SSE and WebSockers not as competing technologies, but rather as 2 tools that may be suitable for different situations. SSE is very lightweight and firewall friendly. It also uni-directional in contrast to WebSockets. Websockets on the other hand are more powerful and open a door for more interesting scenarios and use-cases. 

What I want to say is that I would like to keep both examples around: SSE-based and WebSocket-based. I believe that each of them have a lot of value. If we decide keep then both in the same repo, then we need to restructure code a bit to make good demonstration of every approach in isolation (with some shared code of course). Though it will make an example code a bit more complex for newcomers. An alternative would be to keep these two approaches in 2 different repos. WDYT?

**Comment by trbngr** [comment]
Thanks for the review, Oleg!

I agree with you 100% and to keep it as simple as possible, I think the two repo approach is best. 

I have made a bunch of improvements to this code in my real project that I think benefits this code too. But, I'm going on vacation tomorrow for the next week and I have some obligations to meet before I can leave. No time for OSS.

When I get back, I'll be in touch.

Thanks again,
Chris

**Pull request closed without merging**